File: tests/scripts/script-tests.js
const fs = require('fs');
const vm = require("vm");
const assert = require('chai').assert;
global.importScripts = (...files) => {
  let scripts;

  if (files.length > 0) {
    scripts = files.map(file => {
      //Essentially, we copy pasted from tiny-worker but changed it to use app/scripts/compiler as the source
      return fs.readFileSync("app/scripts/compiler/"+file, "utf8");
    }).join("\n");

    vm.createScript(scripts).runInThisContext();
  }
};
importScripts("includes.js");
fs.writeFileSync("tests/results.txt","");
fs.readdirSync("tests").forEach(result => testScript("tests/"+result));

function testScript(script) {
  if (script.endsWith("results.txt") || !script.endsWith("txt") || script.endsWith("js")) return;
  const code = fs.readFileSync(script, 'utf-8');
  let compile = Compiler.compileWithoutWorker(code, {isLocal: true, isFairAbstraction: true});

  describe(script, function () {
    //Fail.txt should not compile successfully
    if (script.endsWith("fail.txt")) {
      failTxt(compile);
      return;
    }
    it('The script should compile successfully', function () {
      assert(compile.type !== 'error', "Error compiling, Message: " + compile.message);
    });
    const operations = compile.operations;
    if (operations && operations.length > 0) {
      for (let i = 0; i < operations.length; i++) {
        const {operation, process1, process2, result} = operations[i];
        const op = process1 + ' ' + operation + ' ' + process2;
        if (result == "notfound") {
          let msg = "Error processing operation, "+process1.ident + ' ' + operation + ' ' + process2.ident;
          if (!process1.exists) msg+="\n"+process1.ident+" does not exist!";
          if (!process2.exists) msg+="\n"+process2.ident+" does not exist!";
          if (script.endsWith("nonExistantOperation.txt")) {
            continue;
          }
          it("Error processing operation, "+process1.ident + ' ' + operation + ' ' + process2.ident, function () {
            assert(process1.exists, process1.ident+" does not exist");
            assert(process2.exists, process2.ident+" does not exist");
          });
          continue;
        }
        if (script.endsWith("nonExistantOperation.txt")) {
          it('The script should not compile successfully', function () {
            assert(compile.type ==='error', "nonExistantOperations.txt compiled successfully");
          });
        }
        if (script.endsWith("failOperations.txt")) {
          failOperationsTxt(result, op);
          continue;
        }
        it("Operation '"+op+"' should pass", function () {
          assert(result, "'"+op+"' failed");
        });
      }
    }
  })
}
function failOperationsTxt(result, op) {
  if (op == "A ~ B") {
    it('The script should not pass', function () {
      assert(!result, "'"+op+"' passed");
    });
    return;
  }
  it('The script should pass', function () {
    assert(result, "'"+op+"' failed");
  });
}
function failTxt(compile) {
  it('The script should not compile successfully', function () {
    assert(compile.type ==='error', "fail.txt compiled successfully");
  });
}
function combineEdges(edge1,edge2) {
  const java = require("java");
  const baseDir = "lib";
  const dependencies = fs.readdirSync(baseDir);
  //Load java dependancies
  dependencies.forEach(function(dependency){
    java.classpath.push(baseDir + "/" + dependency);
  });
  //Initilize a solver
  const EdgeMerger = java.import('net.modelsolver.EdgeMerger');
  //Solve

  return JSON.parse(new EdgeMerger().mergeEdgesSync(JSON.stringify(edge1),JSON.stringify(edge2)));
}

File: modelchecker/src/test/java/mc/util/expr/ScriptTests.java
package mc.util.expr;

import mc.compiler.OperationResult;
import mc.exceptions.CompilationException;
import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class ScriptTests {
    @Test
    public void testScripts() throws IOException {
//        for (File file:new File("tests").listFiles()) {
//            if (file.getName().endsWith("results.txt") || !file.getName().endsWith("txt")) return;
//            List<OperationResult> operations = Collections.emptyList();
//            boolean failed = false;
//            try {
//                mc.compiler.Compiler compiler = new mc.compiler.Compiler();
//                operations = compiler.compile(String.join("\n",Files.readAllLines(file.toPath()))).getOperationResults();
//                failed = true;
//            } catch (CompilationException e) {
//                e.printStackTrace();
//            }
//            if (shouldFail(file.getName())) {
//                assertTrue("Test script: "+file.getName()+" should not compile!",failed);
//            } else {
//                assertFalse("Test script: "+file.getName()+" should compile!",failed);
//            }
//
//            if (operations.size() > 0) {
//                for (int i = 0; i < operations.size(); i++) {
//                    OperationResult result = operations.get(i);
//                    String op = result.getProcess1().getIdent() + ' ' + result.getOperation() + ' ' + result.getProcess2().getIdent();
//                    if (Objects.equals(result.getResult(), "notfound")) {
//                        if (file.getName().endsWith("nonExistantOperation.txt")) {
//                            continue;
//                        }
//                        System.out.println(op);
//                        assertTrue(result.getProcess1().getIdent()+" does not exist", result.getProcess1().isExists());
//                        assertTrue(result.getProcess2().getIdent()+" does not exist", result.getProcess1().isExists());
//                        continue;
//                    }
//                    if (shouldFailOperations(file.getName(),op)) {
//                        assertFalse("Operation '"+op+"' should fail",Objects.equals(result.getResult(), "false"));
//                    } else {
//                        assertTrue("Operation '"+op+"' should pass",Objects.equals(result.getResult(), "true"));
//                    }
//
//                }
//            }
//        }
    }
    private boolean shouldFail(String fileName) {
        if (fileName.contains("fail.txt")) return true;
        return false;
    }
    private boolean shouldFailOperations(String fileName, String op) {
        if (fileName.contains("failOperations.txt") && Objects.equals(op, "A ~ B")) return true;
        return false;
    }

}
